api/api-environments-v3/src/main/java/com/thoughtworks/go/apiv3/environments/EnvironmentsControllerV3.java

```diff
@@ -0,0 +1,268 @@
+            throw haltBecauseEtagDoesNotMatch("environment", environmentName);
+        }
+
+        if (!StringUtils.equals(environmentName, environmentConfig.name().toString())) {
+            throw haltBecauseRenameOfEntityIsNotSupported("environment");
+        }
+
+        environmentConfigService.updateEnvironment(environmentConfig.name().toString(), environmentConfig, currentUsername(), etagFor(oldEnvironmentConfig), operationResult);
+
+        setEtagHeader(environmentConfig, response);
+        return handleCreateOrUpdateResponse(request, response, environmentConfig, operationResult);
+    }
+
+    public String partialUpdate(Request request, Response response) {
+        JsonReader jsonReader = GsonTransformer.getInstance().jsonReaderFrom(request.body());
+        PatchEnvironmentRequest req = PatchEnvironmentRequestRepresenter.fromJSON(jsonReader);
+        HttpLocalizedOperationResult result = new HttpLocalizedOperationResult();
+
+        String environmentName = request.params("name");
+        EnvironmentConfig environmentConfig = fetchEntityFromConfig(environmentName);
+
+        Optional<EnvironmentVariableConfig> parsingErrors = req.getEnvironmentVariablesToAdd().stream()
+                .filter(envVar -> !envVar.errors().isEmpty())
+                .findFirst();
+
+        if (parsingErrors.isPresent()) {
+            EnvironmentVariablesConfig configs = new EnvironmentVariablesConfig(req.getEnvironmentVariablesToAdd());
+            String errorMessage = MessageJson.create("Error parsing patch request",
+                    writer -> toJSONArray(writer, "environment_variables", configs));
+            response.status(422);
+            return errorMessage;
+        }
+
+        List<String> dummyAgentList = new ArrayList<>();
+        environmentConfigService.patchEnvironment(environmentConfig, req.getPipelineToAdd(), req.getPipelineToRemove(),
+                dummyAgentList, dummyAgentList, req.getEnvironmentVariablesToAdd(),
+                req.getEnvironmentVariablesToRemove(), currentUsername(), result);
+
+        EnvironmentConfig updateConfigElement = fetchEntityFromConfig(environmentName);
+
+        setEtagHeader(updateConfigElement, response);
+        return handleCreateOrUpdateResponse(request, response, updateConfigElement, result);
+    }
+
+    public String remove(Request request, Response response) {
+        String environmentName = request.params("name");
+        HttpLocalizedOperationResult result = new HttpLocalizedOperationResult();
+        EnvironmentConfig environmentConfig = fetchEntityFromConfig(environmentName);
+
+        environmentConfigService.deleteEnvironment(environmentConfig, currentUsername(), result);
+        return handleSimpleMessageResponse(response, result);
+    }
+
+    @Override
+    public String etagFor(EnvironmentConfig entityFromServer) {
+        if (entityFromServer instanceof MergeEnvironmentConfig || entityFromServer instanceof UnknownEnvironmentConfig) {
```

Comment:
shouldn't this logic be under `entityHashingService.md5ForEntity(EnvironmentConfig)`?

Now we have the `md5Hex` computation logic at 2 places. And moreover, the controller shouldn't have the logic of etag computation.

Reply:
Agreed. Moved the logic to the suggested place.